make this work for all of pg

eshellman opened the pull request:
To make this work for the 50K books in Project Gutenberg, I needed to make the following changes:

1. print subtitles in smaller font on a new line.
2. scale fonts down for very long titles.
3. use a font that prints chinese characters

eshellman committed:
make this work for all of pg

1. print subtitles in smaller font on a new line.
2. scale fonts down for very long titles.
3. use a font that prints chinese characters

jenstroeger commented on tenprintcover.py:
```diff
@@ -498,25 +498,43 @@ def drawShape(c, x, y, s):
 
-
+    def scale_font(text, font_name, font_properties):
+        # very dumb heuristic
```
Whole sentences and perhaps a better explanation?

jenstroeger commented on tenprintcover.py:
```diff
@@ -498,25 +498,43 @@ def drawShape(c, x, y, s):
+        # very dumb heuristic
+        (font_size, font_slant, font_weight) = font_properties
+        w = len(text) * font_size
+        print 100 * w
```
Remove debug print.

eshellman replied:
whoops!

jenstroeger commented on tenprintcover.py:
```diff
@@ -498,25 +498,43 @@ def drawShape(c, x, y, s):
-        title_font = cover_image.font("Avenir Next", title_font_properties)
+        subtitle_font_properties = (subtitle_font_size, cairo.FONT_SLANT_NORMAL,
+                                     cairo.FONT_WEIGHT_NORMAL)
+        title_font_properties = scale_font(title, "Arial Unicode MS", title_font_properties)
```
Nitpick... Arial? Really!? 🤔

eshellman replied:
I know, but it was the only typeface on my computer that worked for all the titles. Maybe I'll investigate the google fonts.

jenstroeger replied:
Perhaps [Noto Sans](https://fonts.google.com/specimen/Noto+Sans) would be a safe option, considering it covers the Unicode pages completely. It would also make for a nice addition to specify a font...

jenstroeger commented on tenprintcover.py:
```diff
@@ -525,14 +543,13 @@ def drawText():
     # If any, append the book's subtitle to the title.
-    if subtitle:
-        title += ": " + subtitle
+    #if subtitle:
```
Remove commented code completely. We’ll have it in Git here. Make sure to handle absence of subtitles gracefully.

eshellman commented:
thanks so much for the review, I'll make the fixes.

eshellman committed:
address @jenstroeger suggestions

eshellman committed:
Add font requirement

jenstroeger committed:
Merge branch 'master' into harden-for-gitenberg

jenstroeger committed:
Minor cleanup.

jenstroeger commented on tenprintcover.py:
```diff
@@ -498,25 +498,50 @@ def drawShape(c, x, y, s):
+        for char in text:
+            if ord(char) >= 0x4E00:
+                return 'Noto Sans CJK SC'
+        return 'Noto Sans'
```
Instead of a manual iteration and a hardwired check against the [16b Unihan constant `4E00`](http://unicode.org/cgi-bin/GetUnihanData.pl?codepoint=4e00), I would like to propose using the [regex](https://pypi.python.org/pypi/regex) module which supports Unicode character properties, Unicode [Blocks](https://en.wikipedia.org/wiki/Unicode_character_property#Block) and [Scripts](https://en.wikipedia.org/wiki/Unicode_character_property#Script) in particular:

```python
import regex

# Pre-compile regular expression for later: search for characters covered
# by the Han script; this is easier than specifying the blocks which are
# covered by the script.
_re = regex.compile(r"\p{Script=Han}+")

def select_font(text):
    return "Noto Sans" if _re.search(text) is None else "Noto Sans CJK SC"
```

We’d probably have to extend the script though, to [Hrkt (or Kana)](https://en.wikipedia.org/wiki/Kana) for the Japanese scripts, and to [Kore (Han + Hangul)](https://en.wikipedia.org/wiki/Korean_mixed_script) for the Korean. Also, consider dropping the `+` because the existence of a single character is enough to demand a font switch.

```python
_re = regex.compile(r"(\p{Script=Han}\p{Script=Hrkt}\p{Script=Kore})+")
```

eshellman replied:
Wouldn't object to this change, but testing needs a lot more examples. I suggest creating a separate issue. Over the next month or so, I'll gather use cases.

jenstroeger replied:
If you would like to check this code in as-is, then yes, we should open a high-pri issue and address it soon. Personally, I would prefer using a regex with Unicode Script support and avoid hard-wired constants.

Note though that by looking for Unicode point `4E00` and above, you miss other ranges below (e.g. [CJK Unified Ideographs Extension A, from `3400`..`4DBF`](https://unicode-table.com/en/blocks/cjk-unified-ideographs-extension-a/), and include non-CJK scripts that start at `A000` (the [Yi Syllables](https://unicode-table.com/en/blocks/yi-syllables/)).

If you look at a [linear expansion of the Unicode table](https://unicode-table.com/en/) you'll notice how CJK-related blocks are scattered throughout, which makes the use of the hard-wired constant quite unreliable.

eshellman replied:
I think the [FontTools solution](http://stackoverflow.com/questions/4458696/finding-out-what-characters-a-font-supports#19438403) is the best way to address font selection, but I have too many other fish to fry at the moment to work on it.

jenstroeger replied:
If you refer to our previous conversation and [this comment and code](https://github.com/mgiraldo/tenprintcover-py/pull/4#issuecomment-282867591), then I would agree. Either way, we should open an issue for this.

eshellman commented:
I'm not familiar with that. Seems like better code. test it with title =  "粉妝樓41-50回" (pg 4576)

jenstroeger commented:
@eshellman, Do you have a few Japanese, Chinese (traditional etc), and Korean titles? Your test title works:

```python
>>> import regex
>>> _re = regex.compile(r"[\p{Script=Han}\p{Script=Hrkt}\p{Script=Hangul}]")
>>> def select_font(text):
...     return "Noto Sans" if _re.search(text) is None else "Noto Sans CJK SC"
... 
>>> select_font("粉妝樓41-50回") # Chinese
'Noto Sans CJK SC'
>>> select_font("1987년 10월 29일") # Korean
'Noto Sans CJK SC'
>>> select_font("漢字") # Kanji
'Noto Sans CJK SC'
>>> select_font("Märchen der Gebrüder Grimm") # German
'Noto Sans'

```

Note though, that if we were to use this implementation, we’d create a dependency on the [`regex`](https://pypi.python.org/pypi/regex) module.

eshellman commented:
I didn't notice that the regex you suggest to import is not python's re.

I think that to really do this right would need locale awareness to deal with zh-HK, zh-TW and zh-CN, but that's a bigger project. I'd prefer the simpler 99% solution.

A better use of resources would be to make the font family a parameter. I didn't want to do that before finding a use case.

In a week or so, we'll have deployed this on 50K books; we can look for edge cases then.

jenstroeger commented:
Python’s [locale](https://docs.python.org/3.6/library/locale.html) is an interface to your machine’s [POSIX locale](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap07.html) which _“is the definition of the subset of a user's environment that depends on language and cultural conventions_”.

If you wanted to make use of a locale, I think you’d first have to analyze the title and author text for the language, then switch the locale (using the [setlocale()](https://docs.python.org/3.6/library/locale.html#locale.setlocale) function) to gain access to date and number formatting and so forth. I’m not sure if that’s what you really need.

The problem in the case here is to find a font whose glyphs cover all letters of a given string (the book’s title, subtitle, and author). For that, [Google’s Noto Sans](https://fonts.google.com/specimen/Noto+Sans) or [GNU Unifont](http://unifoundry.com/unifont.html) seem appropriate, but there are perhaps more.

I think what you really need, in addition to a good font, is some code to [find out if the font supports all characters of your text](http://stackoverflow.com/questions/4458696/finding-out-what-characters-a-font-supports), i.e. if the character exists in the font’s `cmap` table. Then run all text through this test and warn if you come across a character for which Noto or Unifont do not provide a glyph—I’m sure the owners of the font would like to know, too.

Personally, I’d use a wide-ranging font and then check the text. I don’t think switching fonts whenever this or that character is in use is really necessary.

jenstroeger commented:
Based on [this Stackoverflow answer](http://stackoverflow.com/questions/4458696/finding-out-what-characters-a-font-supports#19438403) I think something like this would work:

```python
import fontTools.ttLib

# Open and parse the font file.
ttf = fontTools.ttLib.TTFont("/path/to/fonts/NotoSans-Regular.ttf", 0)

# Get a list of Unicode code points of all glyphs in all character maps.
ords = set(_ for keys in [_.cmap.keys() for _ in ttf["cmap"].tables] for _ in keys)

# Check if a character is in that list.
ord("A") in ords
ord("한") in ords
```

The code needs some refinement, and also creates a dependency on the Python [`fonttools`](https://pypi.python.org/pypi/FontTools) package.

eshellman commented:
I found some issues with subtitles because NotoSans metrics are different. Will add to this pr.

jenstroeger commented:
> I found some issues with subtitles because NotoSans metrics are different. Will add to this pr.

Curious... which kind of issues? Similar problems with [Noto Serif](https://fonts.google.com/specimen/Noto+Serif)? Looking forward to the PR update. (As a side note, [this FAQ](https://github.com/googlei18n/noto-fonts/blob/master/FAQ.md#could-you-provide-a-single-font-file-that-covers-every-language-or-at-least-as-many-scripts-as-possible) explains why there is not one single Noto font file which covers all scripts.)

eshellman commented:
title:subtitle = 'A History of Southern Missouri and Northern Arkansas

  Being an Account of the Early Settlements, the Civil War, the Ku-Klux, and Times
  of Peace'

jenstroeger commented:
@eshellman Basically, a title with two successive `\n\n` linefeeds? Humph, now things start to get messy.

Is it possible to run an initial scan over all books to check for something like this, clean it up, and assume that for the book cover generation neither title nor subtitle contain linefeeds? I think it would be better to clean up the books themselves than trying to bend the book cover generation around it.

eshellman commented:
@jenstroeger this the code I'm using to prepare the title

```
title_splitter = re.compile(r'[\r\n:]+', flags=re.M)

    def split_title(self):
        title=self.metadata.get('title', '')
        title = title_splitter.split(title,maxsplit=1)
        return title if len(title)>1 else [title[0],'']
    
    @property
    def subtitle(self):
        return self.split_title()[1]
        
    @property
    def title_no_subtitle(self):
        return self.split_title()[0]

```
so the example is really 
title=A History of Southern Missouri and Northern Arkansas
subtitle=Being an Account of the Early Settlements, the Civil War, the Ku-Klux, and Times
of Peace

The subtitle overlaps the 2- line title

jenstroeger commented:
I see. Have you considered using [Python regex groups](https://docs.python.org/3/library/re.html) and [Match objects](https://docs.python.org/3/library/re.html#match-objects)? That might simplify the above code:
```python
>>> import re
>>> r = re.compile("(?P<title>.*)[\n\r\:]+(?P<subtitle>.*)")
>>> m = r.match("Some title") # Returns None, no match, thus no newlines and no subtitle.
>>> m = r.match("Some title\n\nsubtitle") # Returns a Match object, access title and subtitle by name.
>>> m.group('title')
'Some title'
>>> m.group('subtitle')
'subtitle'
```
In the above regular expression, a title and subtitle can be anything `.*` but it might make sense to constrain this to `[\w\s\d]+`, i.e. words, spaces, and numbers only, plus perhaps some punctuation. That way, you’d be able to identify odd outliers for book titles, and you can perhaps decide to revisit them?

As far as overlapping on the rendered book cover is concerned, if I remember correctly there is code (helper function [`chop()`](https://github.com/mgiraldo/tenprintcover-py/blob/master/tenprintcover.py#L155)) which checks the length of a title and truncates it using an ellipsis. Does that not work correctly?

An alternative to the truncation would be an iterative down-sizing of the font until the text (title/subtitle) fits into the given bounding box. But for _loooong_ titles this might yield tiny cover text :-)

eshellman committed:
rework subtitle placement

the code now counts the lines used for the title. It scales up the
title if short. It scales the subtitle font as well. Also added a link
to  mgiraldo repo.

eshellman committed:
Merge remote-tracking branch 'origin/harden-for-gitenberg' into harden-for-gitenberg

# Conflicts:
#	tenprintcover.py

eshellman commented:
I had implemented one tier of font scaling; I've refined that in the new commits. Gutenberg has all sorts of title craziness, a cover generator is the wrong place to do title hygiene. chop() worked fine, but it was being used way too much. With the current title scaling, it will only come into play for very long titles/subtitles.

The title splitting code is going in gitenberg.metadata.pandata; I've not committed it yet. The cover invocation code is going in gitenberg.book. I will reference this PR when I commit it.

eshellman committed:
add a test

jenstroeger commented on tenprintcover.py:
```diff
@@ -171,12 +172,13 @@ def chop(word):
         self.context.set_antialias(cairo.ANTIALIAS_DEFAULT)
         # Get some font metrics.
-        font_asc, _, font_height, _, _ = self.context.font_extents()
+        font_asc, _, self.font_height, _, _ = self.context.font_extents()
```
In essence, this function now returns the number of lines occupied by the given text (`nlines`), and as a _side-effect_ it stores the used height of the font as a property of the `cover_image` instance in `self.font_height`. (Which will be over-written by the next call.) The caller of this function then uses _both_ values to multiply them only once.

Following the [Zen of Python](https://www.python.org/dev/peps/pep-0020/) that “Explicit is better than implicit” I would propose to avoid implicit side effects and do one of the two:
 1. `return nlines, font_height`, or
 2. `return nlines * font_height`.

This keeps this function side-effect free and the `cover_image` instance immutable.

eshellman replied:
I agree, will change.

jenstroeger commented on .gitignore:
```diff
@@ -1 +1,2 @@
 .DS_Store
+*.pyc
```
Just commit this one to master.

eshellman replied:
I don't have commit privs.

jenstroeger replied:
Oh, this is something for @mgiraldo then.

mgiraldo replied:
now you do :)

eshellman replied:
done

jenstroeger commented on tenprintcover.py:
```diff
@@ -186,17 +188,21 @@ def chop(word):
                     if w_y > height:
-                        return
+                        return nlines
+                    else:
```
Nit: the `else` can be removed and the increment can be un-indented.

jenstroeger commented on tenprintcover.py:
```diff
@@ -498,42 +504,68 @@ def drawShape(c, x, y, s):
-        author_font = cover_image.font("Avenir Next", author_font_properties)
+        author_font = cover_image.font(select_font(author), author_font_properties)
         author_height = (cover_height - cover_width - (cover_height * cover_margin / 100)) * 0.25
-
```
Would you mind leaving empty lines (521, 528) in the code? :-)

eshellman replied:
ok

jenstroeger commented on tenprintcover.py:
```diff
@@ -498,42 +504,68 @@ def drawShape(c, x, y, s):
-        cover_image.text(title, x, y, width, height, fill, title_font)
+        title_lines = cover_image.text(title, x, y, width, height, fill, title_font)
+        if subtitle:
+            y = min( y + cover_image.font_height * title_lines * cover_height, 
```
I think that [`pylint`](https://www.pylint.org/) will nag about spacing here, have you checked? (Always good to run `pylint` over code, I use it as a check-in requirement.)

eshellman replied:
haven't pylinted this, but will remove the extra space.

jenstroeger reviewed (changes requested):
Thank you for all the work! Have you considered using the regular expression to check for CJK characters instead of the the loop?

mgiraldo committed:
Merge branch 'master' into harden-for-gitenberg

eshellman committed:
address review

reduced pylint nags
included font_height in image.text() return
py3 compatibility

jenstroeger commented on tenprintcover.py:
```diff
@@ -32,6 +32,9 @@
 
+PY2 = sys.version_info[0] == 2
+if PY2:
+    FileNotFoundError = IOError
```
Good catch that [Py3’s FileNotFoundError](https://docs.python.org/3/library/exceptions.html#FileNotFoundError) (derived from `OSError`) is absent from Py2 which uses its own [IOError](https://docs.python.org/2/library/exceptions.html#exceptions.IOError) exception. I think this should be a separate pull request because it’s a separate issue.

However, redefining built-in Exception objects like above seems very hacky. Have you tried

```python
except (FileNotFoundError, IOError):
    ...
```

I’ve assigned issue https://github.com/mgiraldo/tenprintcover-py/issues/5 to you.

eshellman replied:
running pylint found this error. I don't think your proposed solutions works, but 

```python
except (IOError, FileNotFoundError):
```

might, depending on the order of evaluation, but that seems hackier.

jenstroeger replied:
Perhaps we should remove these three lines from this PR, as they are discussed in issue https://github.com/mgiraldo/tenprintcover-py/issues/5 separately.

eshellman replied:
I'm running Py2, so for my own work, I'll be leaving this code in.

jenstroeger replied:
As of commit https://github.com/mgiraldo/tenprintcover-py/commit/7cc2fc19b4cee6928c5e012eed15462ffb1e46b2 this code isn’t necessary anymore. If you prefer to keep this change locally, that’s alright, but I don’t think we should check the code into the repo.

jenstroeger commented on tenprintcover.py:
```diff
@@ -305,7 +310,11 @@ def processColors():
         color_seed = int(_map(_clip(counts, 2, 80), 2, 80, 10, 360))
         shape_color = Image.colorHSB(color_seed, base_saturation, base_brightness-(counts % 20))
-        base_color = Image.colorHSB((color_seed + color_distance) % 360, base_saturation, base_brightness)
+        base_color = Image.colorHSB((
```
The second `(` belongs to the arithmetic expression `(color_seed + color_distance) % 360` and should go onto the new line with the expression; it’s not a tuple. Pylint didn’t distinguish that?

eshellman replied:
true...

jenstroeger commented on tenprintcover.py:
```diff
@@ -504,27 +513,28 @@ def drawShape(c, x, y, s):
-        width = len(text) * font_size
-        if width > cover_width * 3:   #This is an empirical, unintelligent, heuristic.
+        w = len(text) * font_size
+        if w > cover_width * 3:   #This is an empirical, unintelligent, heuristic.
```
The `w` was `width` and now changed back to `w`?

eshellman replied:
whoops

jenstroeger commented on tenprintcover.py:
```diff
@@ -504,27 +513,28 @@ def drawShape(c, x, y, s):
     
-    # Return a font appropriate for the text. Uses Noto CJK if text contains letters of
-    # Simplified Chinese, Traditional Chinese, Japanese, and Korean (CJK), otherwise Noto Sans. 
-    # http://www.unicode.org/faq/han_cjk.html
```
This comment was lost with this change?

jenstroeger commented on tenprintcover.py:
```diff
@@ -504,27 +513,28 @@ def drawShape(c, x, y, s):
-    
+
     # Allocate fonts for the title and the author, and draw the text.
+    
```
Remove added newline.

eshellman replied:
I thought you ask for it to be added?

jenstroeger reviewed (changes requested):
Nearly there :)

eshellman committed:
nits

jenstroeger committed:
Update lost comment and keep some original spacing.

jenstroeger committed:
Keep original newlines clean.
